src/Collateral.sol

```diff
@@ -0,0 +1,202 @@
+// Copyright 2023-, Semiotic AI, Inc.
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.13;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {ECDSA} from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import {TAPVerifier} from "./TAPVerifier.sol";
+import {AllocationIDTracker} from "./AllocationIDTracker.sol";
+
+import "forge-std/console.sol";
+
+/**
+ * @title Collateral
+ * @dev This contract allows `senders` to deposit collateral for specific `receivers`,
+ *      which can later be redeemed using Receipt Aggregate Vouchers (`RAV`) signed
+ *      by an authorized `signer`. `Senders` can deposit collateral for `receivers`,
+ *      authorize `signers` to create signed `RAVs`, and withdraw collateral after a
+ *      set `thawingPeriod` number of blocks. `Receivers` can redeem signed `RAVs` to
+ *      claim collateral.
+ * @notice This contract uses the `TAPVerifier` contract for recovering signer addresses
+ *         from `RAVs`.
+ */
+contract Collateral {
+    struct CollateralAccount {
+        uint256 balance; // Total collateral balance for a sender-receiver pair
+        uint256 amountThawing; // Amount of collateral currently being thawed
+        uint256 thawEnd; // Block number at which thawing period ends
+    }
+
+    // Stores how much collateral each sender has deposited for each receiver, as well as thawing information
+    mapping(address => mapping(address => CollateralAccount)) private collateralAccounts;
+    // Map of authorized signers to which sender they are authorized to sign for
+    // key: signer address, value: sender address
+    mapping(address => address) private authorizedSigners;
+
+    // The ERC20 token used for collateral
+    IERC20 public immutable collateralToken;
```

Comment:
Not all ERC20 tokens return a boolean on their transfer functions. These tokens would not be compatible with this version of the contract (there are several instances of `require(collateralToken.transfer(msg.sender, amount));` for example.


We could use OpenZeppelin’s Safe ERC20 helpers OR document the decision with a comment saying that it’s only intended to be used with GRT for example.

Reply:
This is using the openzeppelin ERC20 interface which defines returning a bool for the transfer function. I would assume that any contract implementing that interface would have to adhere to the interface, right?

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/7cc2cbfeb56a00bfe055ebe0d27da902d95ed3d3/contracts/token/ERC20/IERC20.sol#L41

Reply:
What is using OZ's ERC20 interface is the storage variable on the collateral contract but that does not say anything about what the actual target ERC20 contract used for it's implementation. 

You could initialize the collateral contract using the address of an ERC20 contract that does not adhere to OZ's interface, or even to the ERC20 spec. Using OZ's IERC20 on the collateral contract only tells the contract to consider an address as a contract that adheres to the IERC20 spec.

Reply:
Not sure I made a good job at explaining this... perhaps a better way of explaining it is to think the IERC20 here is "casting" an address as "a contract that adheres to the ERC20 interface". You can "cast" any address but that does not mean that the address is an ERC20 contract.

Reply:
It sounds like overall the chosen interface is simply a suggestion to the contract deployer. So if someone wanted to disregard the chosen interface there might not even be a `transfer` function?

Reply:
So in this case it essentially doesn't matter what interface we use or what functions/returns. The deployer could use any address and it will get cast as the interface declared in the contract. It seems like technically they could just send in a random address pointing to nothing and the deploy might still work? I'm not sure if I see why changing this to some other interface would resolve that issue. Are you just saying that avoiding the `require` would likely allow a larger amount of ERC implementations to work since it would work whether or not it returns bool (assuming it always reverts on failure)?

Reply:
> So in this case it essentially doesn't matter what interface we use or what functions/returns. The deployer could use any address and it will get cast as the interface declared in the contract. It seems like technically they could just send in a random address pointing to nothing and the deploy might still work?

Exactly. The IERC20 interface is only describing to this contract how you'll be interacting with the address you pass to it. If it's a random address the deploy will work but it will fail at "runtime" when a function call tries to interact with it.

> I'm not sure if I see why changing this to some other interface would resolve that issue. Are you just saying that avoiding the require would likely allow a larger amount of ERC implementations to work since it would work whether or not it returns bool (assuming it always reverts on failure)?

Exactly! The problem is that not all ERC20's revert on transfer failure and not all return true on success. The SafeERC20 interface is used to normalize this and just forget about it. This article has a nice explanation: https://soliditydeveloper.com/safe-erc20. 


To be honest if you **know** which are the ERC20's you'll be interacting with and you know how they behave then you can acknowledge the risk and not use SafeERC20 at all. Just add a comment in the code to document the decision, for example the [billing contract](https://github.com/edgeandnode/billing-contracts/blob/dfe0f0d9d4a4f1f56e065e2730a38a45df11dbac/contracts/Billing.sol#L20) does not use SafeERC20. [Here](https://github.com/edgeandnode/subscription-payments/blob/main/contract/audits/The%20Graph%20A-1%20_%20Macro%20Audits.pdf) is however another recent audit we had where we didn't specify this and the auditors reported it as a low issue (see L2).

Reply:
Makes sense, thanks for the explanation. I think I'll switch to safe-erc20. I added this as issue #12  for tracking.